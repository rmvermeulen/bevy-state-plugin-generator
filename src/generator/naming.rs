use std::borrow::Cow;

use itertools::Itertools;
use lazy_regex::regex;

use crate::NamingScheme;
use crate::models::{ParentState, StateNode};

pub trait NormalizeStateName {
    fn normalize_state_name(&self) -> String;
}

impl<S: ToString> NormalizeStateName for S {
    fn normalize_state_name(&self) -> String {
        split_by_case(&self.to_string()).join("")
    }
}

pub fn apply_naming_scheme(
    naming_scheme: NamingScheme,
    node: &StateNode,
    parent: Option<&ParentState>,
) -> String {
    let name = node.name().normalize_state_name();
    match naming_scheme {
        NamingScheme::None => name,
        NamingScheme::Short => {
            if let Some(parent) = parent {
                let parent = parent.state_name();
                format!("{parent}{name}").normalize_state_name()
            } else {
                name
            }
        }
        NamingScheme::Full => {
            if let Some(parent) = parent {
                let parent = parent.ancestral_name();
                format!("{parent}{name}").normalize_state_name()
            } else {
                name
            }
        }
    }
}

fn split_by_case<'s>(input: &'s str) -> Vec<Cow<'s, str>> {
    regex!("(\\d+|[A-Z][a-z]*|^[a-z]+)")
        .captures_iter(input)
        .flat_map(|c| {
            c.iter()
                .filter_map(|x| x.map(|m| m.as_str()))
                .dedup()
                .collect_vec()
        })
        .peekable()
        .batching(|it| {
            let mut word = Vec::new();
            while let Some(next_token) = it.peek() {
                if next_token.len() == 1 {
                    word.push(it.next().unwrap());
                } else if word.is_empty() {
                    return Some(it.next().unwrap().into());
                } else {
                    return Some(word.into_iter().collect::<Cow<str>>());
                }
            }
            (!word.is_empty()).then(|| word.into_iter().collect::<Cow<str>>())
        })
        .dedup()
        .collect_vec()
}

#[cfg(test)]
mod test_case_thing {

    use itertools::Itertools;
    use rstest::rstest;
    use speculoos::assert_that;

    use super::split_by_case;

    #[rstest]
    #[case("example", vec!["example"])]
    #[case("exampleState", vec!["example","State"])]
    #[case("Example", vec!["Example"])]
    #[case("ExampleState", vec!["Example", "State"])]
    #[case("StateState", vec!["State"])]
    #[case("ExampleStateState", vec!["Example","State"])]
    #[case("ExampleStateExampleState", vec!["Example","State","Example","State"])]
    #[case("ABCThing", vec!["ABC", "Thing"])]
    #[case("AB12C", vec!["AB", "12", "C"])]
    #[case("ABCThingExample", vec!["ABC", "Thing", "Example"])]
    #[case("AAABBBCCCorp", vec!["AAABBBCC","Corp"])]
    fn test_split_case(#[case] input: &str, #[case] expected: Vec<&str>) {
        assert_that!(split_by_case(input))
            .is_equal_to(expected.into_iter().map(|s| s.into()).collect_vec());
    }
}
