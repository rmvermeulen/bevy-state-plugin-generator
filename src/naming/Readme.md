# naming

Apply the correct `NamingScheme`.

## split at case-boundaries, but maintain inner name

| input   | names   | parts      | notes                                |
| ------- | ------- | ---------- | ------------------------------------ |
| `ABC`   |         | `ABC`      | consecutive capitals are joined      |
| `ABBox` |         | `ABBox`    | lowercase sections are joined to the |
|         |         |            | preceding capital(s)                 |
| `AbBox` |         | `Ab`,`Box` | split at case-boundaries             |
| `AbBox` | `AbBox` | `Ab`,`Box` | do not split a name                  |

## normalization

- always suffix the enum/struct with `State` but not the variants, e.g.
  `Map::Fishing` -> `MapState::Fishing` with sub-state `MapStateFishingState`
  |origin name| result| sub-state

### merge

- if `config.merge=true` remove `State` from the end of names when joining
  them e.g. `Map::Fishing` -> `MapState::Fishing` with sub-state `MapFishingState`

## collisions

Abort on any collisions.
