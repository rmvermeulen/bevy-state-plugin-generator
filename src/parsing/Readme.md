# parsing

Parsing is implemented using [`nom`](crates.io/nom). The result of a successfully
parsed text is the root `Node` which contains any children.

See the [tests](./tests.rs) for detailed examples of parsing behavior.
