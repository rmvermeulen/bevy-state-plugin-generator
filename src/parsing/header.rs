use itertools::Itertools;
use lazy_regex::regex;

use crate::generate::get_package_info;
use crate::{NamingScheme, PluginConfig};

#[derive(Clone, Debug, Default, PartialEq)]
pub(crate) struct TemplateHeader<'a> {
    pub template: Vec<&'a str>,
    pub comments_block: Vec<&'a str>,
    pub info_block: Vec<String>,
}

trait Unquote {
    fn unquoted(&self) -> Self;
}

impl Unquote for &str {
    fn unquoted(&self) -> Self {
        self.strip_prefix("\"")
            .and_then(|s| s.strip_suffix("\""))
            .unwrap_or(self)
    }
}

pub(crate) fn parse_template_header<'a>(
    source: &'a str,
    plugin_config: &mut PluginConfig,
) -> TemplateHeader<'a> {
    let leading_comments = source
        .lines()
        .take_while(|line| line.starts_with("//"))
        .collect_vec();

    let mut info_block = Vec::new();
    info_block.push(format!("generated by {}", get_package_info()));

    macro_rules! warn {
        ($($exprs:expr),*) => {
            let warning = format!($($exprs),*);
            let warning = format!("WARN: {warning}");
            eprintln!("{warning}");
            info_block.push(warning);
        };
    }

    let mut template_source = Vec::new();
    let mut in_template = false;
    let leading_comments = leading_comments
        .iter()
        .filter(|line| {
            if in_template {
                if let Some(line) = line.strip_prefix("//") {
                    template_source.push(line.trim());
                    true
                } else {
                    false
                }
            } else if let Some(captures) =
                regex!(r#"^\s*//\s*bspg:(\w+)\s+(\w+|"\w+\")\s*$"#).captures(line)
            {
                let (_, [name, value]) = captures.extract();
                match name {
                    "root_state_name" => {
                        plugin_config.root_state_name = if value == "None" {
                            None
                        } else {
                            Some(value.unquoted().to_string())
                        };
                    }
                    "naming_scheme" => {
                        if let Some(naming_scheme) = NamingScheme::try_parse(value) {
                            plugin_config.naming_scheme = naming_scheme;
                        } else {
                            warn!("invalid naming scheme '{value}' (expected [none, short, full])");
                        }
                    }
                    _ => {
                        warn!("unknown setting: '{name}'");
                    }
                }
                true
            } else if regex!(r#"^\s*//\s*bspg:\s*$"#).is_match(line) {
                in_template = true;
                true
            } else {
                eprintln!("dropping: '{line:?}'");
                false
            }
        })
        .copied()
        .collect_vec();
    TemplateHeader {
        template: template_source,
        comments_block: leading_comments,
        info_block,
    }
}

#[cfg(test)]
mod tests {
    use bevy_utils::default;
    use indoc::formatdoc;
    use rstest::rstest;
    use speculoos::assert_that;
    use speculoos::prelude::{ContainingIntoIterAssertions, VecAssertions};

    use crate::parsing::header::parse_template_header;
    use crate::{NamingScheme, PluginConfig};

    #[rstest]
    #[case(String::new())]
    #[case::ignore_generated_lines(formatdoc! {"
        // generated by v[CARGO_PKG_VERSION]
        // WARN: some warning
    "})]
    fn test_parse_template_header(#[case] header: String) {
        let mut config: PluginConfig = default();
        let header = parse_template_header(&header, &mut config);
        assert_that!(header.template).is_empty();
        assert_that!(header.comments_block).is_empty();
    }

    #[rstest]
    fn test_parse_template_header_keep_variables() {
        let header = formatdoc! {"
            // generated by v[CARGO_PKG_VERSION]
            // WARN: some warning
            // bspg:some setting
        "};
        let mut config: PluginConfig = default();
        let header = parse_template_header(&header, &mut config);
        assert_that!(header.comments_block).contains("// bspg:some setting");
        assert_that!(header.info_block).contains("WARN: unknown setting: 'some'".to_string());
    }

    #[rstest]
    fn test_parse_template_header_variable_values(
        #[values(NamingScheme::Full, NamingScheme::None, NamingScheme::Short)]
        naming_scheme: NamingScheme,
        #[values(None, Some("MyStateRoot"))] name: Option<&str>,
    ) {
        let mut config: PluginConfig = default();
        let header = formatdoc! {"
            // generated by v[CARGO_PKG_VERSION]
            // bspg:root_state_name {}
            // bspg:naming_scheme   {}
            // bspg:
            ",
            name.map(|name| format!("\"{name}\"")).unwrap_or("None".to_string()),
            naming_scheme
        };
        let _ = parse_template_header(&header, &mut config);
        assert_that!(config.root_state_name).is_equal_to(name.map(String::from));
        assert_that!(config.naming_scheme).is_equal_to(naming_scheme);
    }
}
