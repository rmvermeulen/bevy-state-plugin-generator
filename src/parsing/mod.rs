pub(crate) mod header;
mod parsers;
#[cfg(test)]
mod tests;
mod tokens;

pub use parsers::*;
pub(crate) use tokens::*;

// TODO: investigate "tree-less parsing", see https://www.youtube.com/watch?v=NxiKlnUtyio
