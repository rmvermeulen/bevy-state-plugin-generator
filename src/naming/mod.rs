#[cfg(test)]
mod tests;

use std::collections::VecDeque;

use crate::parsing::ParseNode;
use crate::tree::SubTree;

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum NodeType {
    Singleton,
    List,
    Enum,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct NodeData {
    index: usize,
    parent: Option<usize>,
    node_type: NodeType,
    depth: usize,
    name: String,
}

pub fn flatten_parse_node(root_node: ParseNode<'_>) -> Vec<NodeData> {
    let node_count = root_node.get_tree_size();
    let mut nodes = Vec::with_capacity(node_count);
    let mut todo = VecDeque::from([(root_node, 0, None)]);
    while let Some((parse_node, depth, parent)) = todo.pop_front() {
        let Some(name) = parse_node.name() else {
            continue;
        };
        let node_type = match parse_node {
            ParseNode::Singleton(_) => NodeType::Singleton,
            ParseNode::Enum(_, _) => NodeType::Enum,
            ParseNode::List(_, _) => NodeType::List,
            ParseNode::Comment(_) => {
                unreachable!("Comment has no name")
            }
        };
        let index = nodes.len();
        nodes.push(NodeData {
            node_type,
            index,
            parent,
            depth,
            name: name.to_string(),
        });
        for child in parse_node.children() {
            todo.push_back((child, depth + 1, Some(index)));
        }
    }

    for (i, node) in nodes.iter().enumerate() {
        assert_eq!(node.index, i);
        if let Some(parent) = node.parent {
            assert!(parent <= node.index);
            assert!(node.depth > nodes[parent].depth);
        }
    }

    nodes
}
