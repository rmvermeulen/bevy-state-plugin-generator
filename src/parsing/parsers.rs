use nom::branch::alt;
use nom::bytes::complete::*;
use nom::character::complete::*;
use nom::combinator::recognize;
use nom::multi::many0;
use nom::sequence::*;
use nom::{IResult, Parser};

use crate::parsing::{Comment, Identifier, ParseNode, Token};

pub fn parse_comment(input: &str) -> IResult<&str, ParseNode<'_>> {
    comment(input).map_result(ParseNode::Comment)
}

pub fn comment(input: &str) -> IResult<&str, Comment<'_>> {
    use nom::combinator::eof;

    delimited(skip_to(tag("//")), not_line_ending, alt((eof, line_ending)))
        .parse(input)
        .map_result(|c| c.trim().into())
}

pub fn identifier(input: &str) -> IResult<&str, Identifier<'_>> {
    recognize(pair(
        take_till1(|c: char| !c.is_uppercase()),
        take_while(|c: char| c.is_alphanumeric() || c == '_'),
    ))
    .parse(input)
    .map_result(Identifier::from)
}

fn skip_to<P, T, O, E>(parser: P) -> impl Parser<T, Output = O, Error = E>
where
    T: nom::Input,
    P: Parser<T, Output = O, Error = E>,
    <T as nom::Input>::Item: nom::AsChar,
    E: nom::error::ParseError<T>,
{
    preceded(multispace0, parser)
}

pub fn separator(input: &str) -> IResult<&str, Token> {
    skip_to(tag(","))
        .parse(input)
        .map_result(|_| Token::Separator)
}

pub fn open_enum(input: &str) -> IResult<&str, Token> {
    skip_to(tag("{"))
        .parse(input)
        .map_result(|_| Token::OpenEnum)
}

pub fn close_enum(input: &str) -> IResult<&str, Token> {
    skip_to(tag("}"))
        .parse(input)
        .map_result(|_| Token::CloseEnum)
}

pub fn open_list(input: &str) -> IResult<&str, Token> {
    skip_to(tag("["))
        .parse(input)
        .map_result(|_| Token::OpenList)
}

pub fn close_list(input: &str) -> IResult<&str, Token> {
    skip_to(tag("]"))
        .parse(input)
        .map_result(|_| Token::CloseList)
}

pub trait MapResult<'a, I, O1, O2> {
    fn map_result(self, callback: impl Fn(O1) -> O2) -> IResult<I, O2>;
}

impl<I, O1, O2> MapResult<'_, I, O1, O2> for IResult<I, O1> {
    fn map_result(self, callback: impl Fn(O1) -> O2) -> IResult<I, O2> {
        self.map(|(rest, result)| (rest, callback(result)))
    }
}

pub fn parse_config(input: &str) -> IResult<&str, Vec<ParseNode<'_>>> {
    many0(delimited(many0(separator), parse_node, many0(separator))).parse(input)
}

pub fn parse_node(input: &'_ str) -> IResult<&'_ str, ParseNode<'_>> {
    alt((parse_enum, parse_list, parse_comment, parse_singleton)).parse(input)
}

pub fn parse_singleton(input: &'_ str) -> IResult<&'_ str, ParseNode<'_>> {
    skip_to(identifier)
        .parse(input)
        .map_result(ParseNode::singleton)
}

pub fn parse_enum(input: &str) -> IResult<&str, ParseNode<'_>> {
    let (input, name) = terminated(skip_to(identifier), skip_to(open_enum)).parse(input)?;
    let (input, children) = skip_to(parse_elements_until(close_enum)).parse(input)?;
    Ok((input, ParseNode::Enum(name, children)))
}

pub fn parse_list(input: &'_ str) -> IResult<&'_ str, ParseNode<'_>> {
    let (input, name) = skip_to(identifier).parse(input)?;
    let (input, children) =
        skip_to(preceded(open_list, parse_elements_until(close_list))).parse(input)?;
    Ok((input, ParseNode::List(name, children)))
}

pub fn parse_elements_until<'a>(
    until: impl Fn(&'a str) -> IResult<&'a str, Token> + Copy,
) -> impl Fn(&'a str) -> IResult<&'a str, Vec<ParseNode<'a>>> {
    move |input: &'a str| {
        delimited(
            // ignore any leading whitespace and commas
            many0(separator),
            // 0 or more elements, ignoring whitespace and commas
            many0(alt((terminated(parse_node, many0(separator)),))),
            // then expect the closing token
            until,
        )
        .parse(input)
    }
}
