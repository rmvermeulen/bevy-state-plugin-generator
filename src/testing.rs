pub use insta::*;
pub use rstest::*;
pub use speculoos::prelude::*;

pub use crate::set_snapshot_suffix;

#[macro_export]
#[allow(missing_docs)]
macro_rules! set_snapshot_suffix {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_suffix(format!($($expr,)*));
        let _guard = settings.bind_to_scope();
    }
}

pub mod node_data {
    use bevy_utils::default;

    use super::*;
    use crate::processing::NodeData;

    fn node<S: ToString>(name: S, index: usize, depth: usize, parent: Option<usize>) -> NodeData {
        let name = name.to_string();
        NodeData {
            index,
            parent,
            resolved_name: Some(name.clone()),
            name,
            depth,
            ..default()
        }
    }

    #[fixture]
    pub fn single_node() -> Vec<NodeData> {
        vec![node("RootState", 0, 0, None)]
    }

    #[fixture]
    pub fn nested_example() -> Vec<NodeData> {
        vec![
            node("RootState", 0, 0, None),
            node("Menu", 1, 1, Some(0)),
            node("Options", 2, 2, Some(1)),
            node("Audio", 3, 3, Some(2)),
            node("Video", 4, 3, Some(2)),
        ]
    }
}

pub mod parse_node {
    use super::*;
    use crate::parsing::Node;

    #[fixture]
    pub fn duplicate_names() -> Node<'static> {
        Node::list("Root", [Node::singleton("A"), Node::singleton("A")])
    }

    #[fixture]
    pub fn overlapping_names() -> Node<'static> {
        Node::list(
            "Root",
            [
                Node::enumeration("Root", [Node::singleton("A")]),
                Node::singleton("A"),
            ],
        )
    }

    #[fixture]
    pub fn enum_root_a() -> Node<'static> {
        Node::enumeration("Root", [Node::singleton("A")])
    }

    #[fixture]
    pub fn enum_root_ab() -> Node<'static> {
        Node::enumeration("Root", [Node::singleton("A"), Node::singleton("B")])
    }

    #[fixture]
    pub fn enum_root_a_b() -> Node<'static> {
        Node::enumeration("Root", [Node::enumeration("A", [Node::singleton("B")])])
    }

    #[fixture]
    pub fn enum_root_a_b_up_c() -> Node<'static> {
        Node::enumeration(
            "Root",
            [
                Node::enumeration("A", [Node::singleton("B")]),
                Node::singleton("C"),
            ],
        )
    }

    #[fixture]
    pub fn list_root_a() -> Node<'static> {
        Node::list("Root", [Node::singleton("A")])
    }

    #[fixture]
    pub fn list_root_ab() -> Node<'static> {
        Node::list("Root", [Node::singleton("A"), Node::singleton("B")])
    }

    #[fixture]
    pub fn list_root_a_b() -> Node<'static> {
        Node::list("Root", [Node::list("A", [Node::singleton("B")])])
    }

    #[fixture]
    pub fn list_root_a_b_up_c() -> Node<'static> {
        Node::list(
            "Root",
            [
                Node::list("A", [Node::singleton("B")]),
                Node::singleton("C"),
            ],
        )
    }

    #[fixture]
    pub fn nested_example() -> Node<'static> {
        Node::enumeration(
            "Menu",
            [
                Node::singleton("Main"),
                Node::comment("these are the options"),
                Node::enumeration(
                    "Options",
                    [
                        Node::singleton("Graphics"),
                        Node::singleton("Audio"),
                        Node::singleton("Gameplay"),
                    ],
                ),
                Node::enumeration("Game", [Node::singleton("Save"), Node::singleton("Load")]),
            ],
        )
    }
}
