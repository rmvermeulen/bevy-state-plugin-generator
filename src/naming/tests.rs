use crate::naming::{NodeData, NodeType, flatten_parse_node};
use crate::parsing::ParseNode;
use crate::testing::*;

#[rstest]
fn test_flatten_parse_node_singleton() {
    let results = flatten_parse_node(ParseNode::singleton("A"));
    assert_that!(results).is_equal_to(vec![NodeData {
        index: 0,
        parent: None,
        node_type: NodeType::Singleton,
        depth: 0,
        name: "A".to_string(),
    }]);
}

#[rstest]
#[case(parse_node::enum_root_a(), "enum_root_a")]
#[case(parse_node::enum_root_ab(), "enum_root_ab")]
#[case(parse_node::enum_root_a_b(), "enum_root_a_b")]
#[case(parse_node::enum_root_a_b_up_c(), "enum_root_a_b_up_c")]
fn test_flatten_parse_node_enums(#[case] node: ParseNode, #[case] name: &str) {
    set_snapshot_suffix!("{name}");
    assert_debug_snapshot!(flatten_parse_node(node));
}

#[rstest]
#[case(parse_node::list_root_a(), "list_root_a")]
#[case(parse_node::list_root_ab(), "list_root_ab")]
#[case(parse_node::list_root_a_b(), "list_root_a_b")]
#[case(parse_node::list_root_a_b_up_c(), "list_root_a_b_up_c")]
fn test_flatten_parse_node_lists(#[case] node: ParseNode, #[case] name: &str) {
    set_snapshot_suffix!("{name}");
    assert_debug_snapshot!(flatten_parse_node(node));
}
