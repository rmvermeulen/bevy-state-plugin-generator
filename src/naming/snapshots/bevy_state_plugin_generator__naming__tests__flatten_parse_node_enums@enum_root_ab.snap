---
source: src/naming/tests.rs
expression: flatten_parse_node(node)
---
[
    NodeData {
        index: 0,
        parent: None,
        node_type: Enum,
        depth: 0,
        name: "Root",
    },
    NodeData {
        index: 1,
        parent: Some(
            0,
        ),
        node_type: Singleton,
        depth: 1,
        name: "A",
    },
    NodeData {
        index: 2,
        parent: Some(
            0,
        ),
        node_type: Singleton,
        depth: 1,
        name: "B",
    },
]
